cadc-rest/src/main/java/ca/nrc/cadc/rest/RestServlet.java

```diff
@@ -296,8 +306,15 @@ protected void doit(HttpServletRequest request, HttpServletResponse response,  C
             action.setSyncInput(in);
             action.setSyncOutput(out);
             action.setLogInfo(logInfo);
+            
+            setAuthenticateHeaders(subject, out);
 
             doit(subject, action);
+        } catch (NotAuthenticatedException ex) {
+            logInfo.setSuccess(true);
+            logInfo.setMessage(ex.getMessage());
```

Comment:
The `setFailedAuthenticationHeader` method looks like it only handles OAuth2 errors (from the `getOauth2Error() = null` bit). I would try to include an error attr in an ivoa challenge in the same style; I'm sure that'll be acceptable to others.

Shouldn't all challenges be re-iterated when authentication fails? Might have to move the bulk of  `setAuthenticateHeaders(Subject,SyncOutput)` into a method that can be used in both scenarios so you can blend in the error message with the challenge that was attempted and still include the others. Maybe just `setAuthenticateHeaders(SyncOutput,Subject,NotAuthenticatedException)` and it's called with a null arg either way?

Reply:
Mentioned in the comment below but will again: it should work with both types of tokens and others we add.  The validator which throws them doesn't care what type they are.

Regarding setting all challenges even on failure:  hmm, not sure about that.  I thought it would be confusing because you'd have to find the header with the error message to diagnose your current problem.  Maybe we should discuss more tomorrow because I think it could go either way.